### DISCUSSION TITLE
Accept only Bearer token as auth in Swagger docs

### DISCUSSION THREAD

**Discussion by mayukhnair** [fastapi-users/fastapi-users#973] [Q&A] [answered]
My transport layer is Bearer and my Strategy is Database. As a result, the `Authorise` button in Swagger docs requires the username and password before one can test out any API calls. 

![image](https://user-images.githubusercontent.com/6792796/166632981-331b4241-e05b-4e9e-a120-5a68fddfa192.png)

I want any and all users to be able to test the API calls within Swagger *only* by providing their Bearer token, not their user and password. Is there any way possible to do the same?

**Comment by frankie567** [comment, marked as answer]
Hi @mayukhnair!

How authentication is rendered inside Swagger is controlled by the security scheme provided by FastAPI. For convenience, we use [`OAuth2PasswordBearer`](https://github.com/tiangolo/fastapi/blob/146f57b8f70c5757dc20edc716dba1b96936a8d6/fastapi/security/oauth2.py#L140-L171) by default on this transport.

The solution in this case could be to use [`HTTPBearer`](https://github.com/tiangolo/fastapi/blob/146f57b8f70c5757dc20edc716dba1b96936a8d6/fastapi/security/http.py#L100-L133), which will render with just a header to fill. However, the problem when using this scheme is that it returns an [`HTTPAuthorizationCredentials` object](https://github.com/tiangolo/fastapi/blob/2aaac141ddb79d1f9071f8e02d2b2097f3d4589f/fastapi/security/http.py#L113-L133), which contains both the scheme and the token. On the other hand, [FastAPI Users Authenticator expects a plain string](https://github.com/fastapi-users/fastapi-users/blob/b7734fc8b05b10507bc7f29afde1e15956fefe9d/fastapi_users/authentication/authenticator.py#L165-L172). So we need to overload `HTTPBearer` to manually explode the `HTTPAuthorizationCredentials` object:

```py
class HTTPBearerTokenOnly(HTTPBearer):
    async def __call__(self, request: Request) -> Optional[str]:  # type: ignore
        credentials = await super().__call__(request)
        if credentials:
            return credentials.credentials
        return None


class SimpleBearerTransport(BearerTransport):
    def __init__(self):
        self.scheme = HTTPBearerTokenOnly()  # type: ignore
```

The `type: ignore` are here to prevent a complains from type-checkers, because we are changing the types of the parent classes.

You can then use it normally:

```py
bearer_transport = SimpleBearerTransport()

def get_jwt_strategy() -> JWTStrategy:
    return JWTStrategy(secret=SECRET, lifetime_seconds=3600)

auth_backend = AuthenticationBackend(
    name="jwt",
    transport=bearer_transport,
    get_strategy=get_jwt_strategy,
)
```

EDIT: Fixed following discussion in #987.

**Reply by mayukhnair** [reply]
Update: This hack may cause 401 Unauthorized errors. @frankie567 has identified and offered a solution here: https://github.com/fastapi-users/fastapi-users/discussions/987#discussioncomment-2721825

I can confirm this works.
